File: python3/n-ary_tree_level_order_traversal.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        if not root:
            return []

        ans = []
        stack = [root]

        while stack:
            temp = []
            line = []
            for node in stack:
                line.append(node.val)

                if node.children:
                    temp += node.children

            ans.append(line)
            stack = temp

        return ans

File: golang/n-ary_tree_level_order_traversal.go
package main

import "golang/model"

/*
Space	: O(n)
Time	: O(n)
*/

 func levelOrderNary(root *model.Node) [][]int {
    if root == nil {
        return [][]int{}
    }
    
    var ans [][]int
    stack := []*model.Node{root}

    for len(stack) > 0 {
        var temp []*model.Node
        var line []int
        
        for _, node := range stack {
            line = append(line, node.Val)
            
            if node.Children != nil {
                temp = append(temp, node.Children...)
            }
        }
        
        ans = append(ans, line)
        stack = temp
    }
    
    return ans
}
